sunpy/data/sample.py

```diff
@@ -1,35 +1,70 @@
 """
-This module provides the following sample data files.  These files are
-downloaded when this module is imported for the first time.
+This module provides the following sample data files.  When a sample shortname
+is accessed, the corresponding file is downloaded if needed.  All files can be
+downloaded by calling :func:`~sunpy.data.sample.download_all`.
 
+Summary variables
+-----------------
+.. list-table::
+   :widths: auto
+
+   * - ``file_dict``
+     - Dictionary of all sample shortnames and, if downloaded, corresponding
+       file locations on disk (otherwise, ``None``)
+   * - ``file_list``
+     - List of disk locations for sample data files that have been downloaded
+
+Sample shortnames
+-----------------
 .. list-table::
    :widths: auto
    :header-rows: 1
 
-   * - Variable name
+   * - Sample shortname
      - Name of downloaded file
 """
-import sys
-from pathlib import Path
+# download_sample_data is deprecated and not used here,
+# but during deprecation period want to keep it in this namespace
+# for backwards compatability. noqa stops it being removed by
+# pre-commit as an unused import
+from ._sample import download_sample_data  # noqa
+from ._sample import _SAMPLE_DATA, _get_sample_files
+
+# Add a table row to the module docstring for each sample file
+for _keyname, _filename in sorted(_SAMPLE_DATA.items()):
+    __doc__ += f'   * - ``{_keyname}``\n     - {_filename}\n'
+
+
+# file_dict and file_list are not normal variables; see __getattr__() below
+__all__ = list(sorted(_SAMPLE_DATA.keys())) + ['download_all', 'file_dict', 'file_list']
 
-from ._sample import _SAMPLE_FILES, download_sample_data
 
-files = download_sample_data()
+# See PEP 562 (https://peps.python.org/pep-0562/) for module-level __dir__()
+def __dir__():
+    return __all__
 
-file_dict = {}
-for f in files:
-    name = Path(f).name
-    _key = _SAMPLE_FILES.get(name, None)
-    if _key:
-        setattr(sys.modules[__name__], _key, str(f))
-        file_dict.update({_key: f})
 
-# Sort the entries in the dictionary
-file_dict = dict(sorted(file_dict.items()))
+# See PEP 562 (https://peps.python.org/pep-0562/) for module-level __getattr__()
+def __getattr__(name):
+    if name in _SAMPLE_DATA:
+        return _get_sample_files([_SAMPLE_DATA[name]])[0]
+    elif name == 'file_dict':
+        return {k: _get_sample_files([_SAMPLE_DATA[k]], no_download=True)[0]
+                for k in _SAMPLE_DATA.keys()}
```

Comment:
```suggestion
        return {k: _get_sample_files([v], no_download=True)[0]
                for k,v in _SAMPLE_DATA.items()}
```

Reply:
Whoops, yeah, my additional contortion is no longer necessary after a recent edit

Reply:
Actually, I've now decided that I want sorting back.  This is now changed to:
```python
        return dict(sorted(zip(_SAMPLE_DATA.keys(),
                               _get_sample_files(_SAMPLE_DATA.values(), no_download=True))))
```